relay-server/src/services/metrics/aggregator.rs

```diff
@@ -158,90 +135,100 @@ impl AggregatorService {
     /// to the receiver to send the [`MergeBuckets`] message back if buckets could not be flushed
     /// and we require another re-try.
     ///
-    /// If `force` is true, flush all buckets unconditionally and do not attempt to merge back.
-    fn try_flush(&mut self) {
-        let force_flush = matches!(&self.state, AggregatorState::ShuttingDown);
-
-        let flush_decision = |project_key| {
-            let project = self.project_cache.get(project_key);
-
-            let project_info = match project.state() {
-                ProjectState::Enabled(project_info) => project_info,
-                ProjectState::Disabled => return FlushDecision::Drop,
-                // Delay the flush, project is not yet ready.
-                //
-                // Querying the project will make sure it is eventually fetched.
-                ProjectState::Pending => {
-                    relay_statsd::metric!(
-                        counter(RelayCounters::ProjectStateFlushMetricsNoProject) += 1
-                    );
-                    return FlushDecision::Delay;
-                }
-            };
-
-            let Some(scoping) = project_info.scoping(project_key) else {
-                // This should never happen, at this point we should always have a valid
-                // project with the necessary information to construct a scoping.
-                //
-                // Ideally we enforce this through the type system in the future.
-                relay_log::error!(
-                    tags.project_key = project_key.as_str(),
-                    "dropping buckets because of missing scope",
-                );
-                return FlushDecision::Drop;
-            };
+    /// Returns when the next flush should be attempted.
+    fn try_flush(&mut self) -> Duration {
+        if cfg!(test) {
+            // Tests are running in a single thread / current thread runtime,
+            // which is required for 'fast-forwarding' and `block_in_place`
+            // requires a multi threaded runtime. Relay always requires a multi
+            // threaded runtime.
+            self.do_try_flush()
+        } else {
+            tokio::task::block_in_place(|| self.do_try_flush())
```

Comment:
Might not need this anymore, but the previous aggregators' max on the flush was up to a 200ms. Since we run in a tokio task here, we're blocking all queued other tasks on this tokio worker for the duration of the call. With `block_in_place` we can tell the runtime to clear it's local queue before running the closure. This in theory should bring down p99+ latencies.